File: packages/ios-sdk/Sources/GateFlowKit/Experiment/BucketEngine.swift
import Foundation

/// Total number of buckets (0–9999), supports 0.1% traffic allocation granularity.
public let totalBuckets = 10000

/// Default salt for hash computation.
public let defaultSalt = "victor"

/// Specification for a variant in a bucket computation.
public struct VariantSpec {
    public let variantKey: String
    public let bucketStart: Int
    public let bucketEnd: Int

    public init(variantKey: String, bucketStart: Int, bucketEnd: Int) {
        self.variantKey = variantKey
        self.bucketStart = bucketStart
        self.bucketEnd = bucketEnd
    }
}

/// Specification for an experiment in a bucket computation.
public struct ExperimentSpec {
    public let expId: String
    public let layerId: String
    public let salt: String
    public let bucketStart: Int
    public let bucketEnd: Int
    public let variants: [VariantSpec]

    public init(expId: String, layerId: String, salt: String, bucketStart: Int, bucketEnd: Int, variants: [VariantSpec]) {
        self.expId = expId
        self.layerId = layerId
        self.salt = salt
        self.bucketStart = bucketStart
        self.bucketEnd = bucketEnd
        self.variants = variants
    }
}

/// Result of a bucket computation for a single experiment.
public struct BucketResult: Equatable {
    public let userId: String
    public let experimentKey: String
    public let bucket: Int
    public let variant: String?
    public let layerId: String?
    public let hit: Bool

    public init(userId: String, experimentKey: String, bucket: Int, variant: String?, layerId: String?, hit: Bool) {
        self.userId = userId
        self.experimentKey = experimentKey
        self.bucket = bucket
        self.variant = variant
        self.layerId = layerId
        self.hit = hit
    }

    public static func hit(userId: String, experimentKey: String, bucket: Int, variant: String, layerId: String) -> BucketResult {
        BucketResult(userId: userId, experimentKey: experimentKey, bucket: bucket, variant: variant, layerId: layerId, hit: true)
    }

    public static func notHit(userId: String, experimentKey: String, bucket: Int) -> BucketResult {
        BucketResult(userId: userId, experimentKey: experimentKey, bucket: bucket, variant: nil, layerId: nil, hit: false)
    }
}

/// Core bucketing engine. Pure computation, no side effects.
public enum BucketEngine {

    /// Compute the bucket number (0–9999) for a user in a specific layer.
    public static func computeBucket(userId: String, layerId: String, salt: String) -> Int {
        let hashInput = "\(userId)#\(layerId)#\(salt)"
        let hash = MurmurHash3.hash32(hashInput)
        return Int(abs(hash)) % totalBuckets
    }

    /// Find the variant key for a given bucket number.
    public static func findVariant(bucket: Int, variantSpecs: [VariantSpec]) -> String? {
        for spec in variantSpecs {
            if bucket >= spec.bucketStart && bucket <= spec.bucketEnd {
                return spec.variantKey
            }
        }
        return nil
    }

    /// Check if a bucket falls within an experiment's traffic range.
    public static func isInExperiment(bucket: Int, bucketStart: Int, bucketEnd: Int) -> Bool {
        bucket >= bucketStart && bucket <= bucketEnd
    }

    /// Compute the full bucket result for a single experiment.
    public static func computeBucketResult(userId: String, experiment: ExperimentSpec) -> BucketResult {
        let bucket = computeBucket(userId: userId, layerId: experiment.layerId, salt: experiment.salt)

        guard isInExperiment(bucket: bucket, bucketStart: experiment.bucketStart, bucketEnd: experiment.bucketEnd) else {
            return .notHit(userId: userId, experimentKey: experiment.expId, bucket: bucket)
        }

        guard let variant = findVariant(bucket: bucket, variantSpecs: experiment.variants) else {
            return .notHit(userId: userId, experimentKey: experiment.expId, bucket: bucket)
        }

        return .hit(userId: userId, experimentKey: experiment.expId, bucket: bucket, variant: variant, layerId: experiment.layerId)
    }

    /// Compute bucket results for all experiments at once.
    public static func computeAllBucketResults(userId: String, experiments: [ExperimentSpec]) -> [BucketResult] {
        experiments.map { computeBucketResult(userId: userId, experiment: $0) }
    }
}

File: packages/android-sdk/gateflow/src/main/java/com/gateflow/sdk/experiment/BucketEngine.kt
package com.gateflow.sdk.experiment

import com.gateflow.sdk.misc.MurmurHash3

const val TOTAL_BUCKETS = 10000
const val DEFAULT_SALT = "victor"

/** Specification for a variant in a bucket computation. */
data class VariantSpec(
    val variantKey: String,
    val bucketStart: Int,
    val bucketEnd: Int,
)

/** Specification for an experiment in a bucket computation. */
data class ExperimentSpec(
    val expId: String,
    val layerId: String,
    val salt: String,
    val bucketStart: Int,
    val bucketEnd: Int,
    val variants: List<VariantSpec>,
)

/** Result of a bucket computation for a single experiment. */
data class BucketResult(
    val userId: String,
    val experimentKey: String,
    val bucket: Int,
    val variant: String? = null,
    val layerId: String? = null,
    val hit: Boolean = false,
) {
    companion object {
        fun hit(userId: String, experimentKey: String, bucket: Int, variant: String, layerId: String) =
            BucketResult(userId, experimentKey, bucket, variant, layerId, true)

        fun notHit(userId: String, experimentKey: String, bucket: Int) =
            BucketResult(userId, experimentKey, bucket, hit = false)
    }
}

/** Core bucketing engine. Pure computation, no side effects. */
object BucketEngine {

    /** Compute the bucket number (0–9999) for a user in a specific layer. */
    @JvmStatic
    fun computeBucket(userId: String, layerId: String, salt: String): Int {
        val hashInput = "$userId#$layerId#$salt"
        val hash = MurmurHash3.hash32(hashInput)
        // Use bitmask instead of abs() to avoid Integer.MIN_VALUE overflow
        return (hash and Int.MAX_VALUE) % TOTAL_BUCKETS
    }

    /** Find the variant key for a given bucket number. */
    @JvmStatic
    fun findVariant(bucket: Int, variantSpecs: List<VariantSpec>): String? {
        for (spec in variantSpecs) {
            if (bucket in spec.bucketStart..spec.bucketEnd) {
                return spec.variantKey
            }
        }
        return null
    }

    /** Check if a bucket falls within an experiment's traffic range. */
    @JvmStatic
    fun isInExperiment(bucket: Int, bucketStart: Int, bucketEnd: Int): Boolean =
        bucket in bucketStart..bucketEnd

    /** Compute the full bucket result for a single experiment. */
    @JvmStatic
    fun computeBucketResult(userId: String, experiment: ExperimentSpec): BucketResult {
        val bucket = computeBucket(userId, experiment.layerId, experiment.salt)

        if (!isInExperiment(bucket, experiment.bucketStart, experiment.bucketEnd)) {
            return BucketResult.notHit(userId, experiment.expId, bucket)
        }

        val variant = findVariant(bucket, experiment.variants)
            ?: return BucketResult.notHit(userId, experiment.expId, bucket)

        return BucketResult.hit(userId, experiment.expId, bucket, variant, experiment.layerId)
    }

    /** Compute bucket results for all experiments at once. */
    @JvmStatic
    fun computeAllBucketResults(userId: String, experiments: List<ExperimentSpec>): List<BucketResult> =
        experiments.map { computeBucketResult(userId, it) }
}
